### ISSUE TITLE
Handle `connection limit` gracefully

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#3175]
## Problem

Having:

```sql
alter user postgrest_test_authenticator with connection limit 2;
```

```
PGRST_DB_CHANNEL_ENABLED=false PGRST_DB_POOL=3 postgrest-with-postgresql-16 postgrest-run
```

Makes PostgREST reset its schema cache and pool when exceeding the connection limit:

```
[nix-shell:~/Projects/postgrest]$ parallel-curl 3 localhost:3000/projects 
```

```
19/Jan/2024:13:28:40 -0500: {"code":"PGRST000","details":"connection to server on socket \"/run/user/1000/postgrest/postgrest-with-postgresql-16-tWK/socket/.s.PGSQL.5432\" failed: FATAL:
  too many connections for role \"postgrest_test_authenticator\"\n","hint":null,"message":"Database connection error. Retrying the connection."}
19/Jan/2024:13:28:40 -0500: Attempting to connect to the database...
127.0.0.1 - postgrest_test_anonymous [19/Jan/2024:13:28:41 -0500] "GET /projects HTTP/1.1" 503 - "" "curl/7.88.1"
19/Jan/2024:13:28:40 -0500: Connection successful
19/Jan/2024:13:28:40 -0500: Config reloaded
19/Jan/2024:13:28:41 -0500: Schema cache loaded
```

It still works normally but there's unnecessary work done.

## Solution

Log the `too many connections` error like above but don't reset the schema cache or pool. 

## Alternatives

Limit the `db-pool` size based on CONNECTION LIMIT?

**Comment by laurenceisla** [comment]
> I understand that are reloaded but I can't seem to find any place where they are reset.

I think that reloaded = reset in this case. There is no "default" value for the schema cache or connection to be reset to AFAIK.

**Comment by taimoorzaeem** [comment]
To me, this seems rather complicated for cases with JWT authentication, because in that case we would need to query the active connections for the user after authenticating and extracting the role first, right?

**Comment by laurenceisla** [comment]
> because in that case we would need to query the active connections for the user after authenticating and extracting the role first, right?

I don't think that's necessary. We don't need to avoid the error, "just" catch it, verify that it's a case of "too many connections for role ..." and not reload the schema cache in those cases, irrespective of the role (I quoted "just" because I didn't check the code, not sure if it's trivial or not). I think most of the `FATAL` connection errors reload the schema, this one would be the exception.

**Comment by taimoorzaeem** [comment]
> ```$ postgrest-parallel-curl 3 localhost:3000/... ```

I am finding it challenging to reproduce this with io test fixtures. Maybe I am missing something?

**Comment by laurenceisla** [comment]
`postgrest-parallel-curl` is a do-while of curl commands:

https://github.com/PostgREST/postgrest/blob/5e0da40ac49f7e742b0727af0929080f7111fea8/nix/tools/devTools.nix#L312-L322

I assume you'd need to do something similar in the IO tests, or find another way to make three requests simultaneously ~~(I think `request()` is synchronous, so doing three requests in a row may work).~~

**Comment by laurenceisla** [comment]
> ~~(I think request() is synchronous, so doing three requests in a row may work).~~

Nah, I'm wrong here. There's an IO test that makes use of parallel requests, you can use it as reference:

https://github.com/PostgREST/postgrest/blob/5e0da40ac49f7e742b0727af0929080f7111fea8/test/io/test_io.py#L694-L695

**Comment by taimoorzaeem** [comment]
My apologies in advance for spamming this thread with unrelated development issue.

I am noticing some difference in `io-test` fixtures as compared to `spec-test` fixtures. More specifically with the same test, I am getting table not found error with io-test fixtures, which is confusing.

```sh
PGRST_DB_CHANNEL_ENABLED=false PGRST_DB_POOL=3 postgrest-with-postgresql-16 -f test/io/fixtures.sql \
postgrest-run
```

```
$ postgrest-parallel-curl 3 localhost:3000/projects 
```

```
{"code":"PGRST205","details":null,"hint":null,"message":"Could not find the table 'test.projects' in the schema cache"}
{"code":"PGRST205","details":null,"hint":null,"message":"Could not find the table 'test.projects' in the schema cache"}
{"code":"PGRST205","details":null,"hint":null,"message":"Could not find the table 'test.projects' in the schema cache"}
```

The `projects` table is in the fixtures. Just wondering why the inconsistency? Is this expected? Maybe I am missing something.

**Comment by wolfgangwalther** [comment]
> The `projects` table is in the fixtures. Just wondering why the inconsistency? Is this expected? Maybe I am missing something.

I think the table `projects` is in the `public` schema according to the fixtures, not the `test` schema.

**Comment by taimoorzaeem** [comment]
> I think the table `projects` is in the `public` schema according to the fixtures, not the `test` schema.

True. but we have `public` default for `db-schemas` and `db-extra-search-path`, so shouldn't it search the table in there too?

**Comment by wolfgangwalther** [comment]
`postgrest-with-...` sets `PGRST_DB_SCHEMAS=test`, so the default for `db-schemas` doesn't apply. `db-extra-search-path` doesn't apply for the endpoints at all.

**Comment by steve-chavez** [comment]
> Limit the db-pool size based on CONNECTION LIMIT?
> i.e. silently? -1
> I sure want to know by looking at the logs, that somehow my configuration is off.
> https://github.com/PostgREST/postgrest/issues/3175#issuecomment-1902119595

There's the idea of treating CONNECTION LIMIT as an in-db config of db-pool in https://github.com/PostgREST/postgrest/issues/4663#issuecomment-4276067761, but it does raise the above problem. It could be confusing as we would apply it silently. Although we could log that we made this decision, something like `"detecting CONNECTION LIMIT on authenticator, setting pool max size as its value"`.

**Comment by wolfgangwalther** [comment]
Funny that I proposed the same thing I had -1ed before ;)

I think the difference in thinking for me between the two issues was, that I would intend to effectively deprecate setting `db-pool` via config in https://github.com/PostgREST/postgrest/issues/4663#issuecomment-4276067761. It would still be supported for backwards-compat, maybe for a while or so, but the *primary* way of setting the pool size should be limiting the number of connections.

If that's documented, there is no silent-behind-your-back-change of your config. During the deprecation phase, we could log something like "db-pool is deprecated and doesn't have an effect, because CONNECTION LIMIT is set" or so.
